### PR TITLE
handle_state_normal: do not update topology of removed endpoint

### LINKED ISSUE #11925: After changing node ip and keeping its host_id, the old endpoint still appears in system.peers
Seen consistently in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-debug/84/testReport/update_cluster_layout_tests/TestUpdateClusterLayout/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split012___test_change_node_ip/
For example:
```
    def check_peer_and_local_table(self, nodes):
        peers_and_local = []
        for node in nodes:
            status = nodetool_status(node)
            logger.debug("nodetool status from {}: {}".format(node.name, status))
            s = self.patient_exclusive_cql_connection(node)
            peers = rows_to_list(s.execute("SELECT host_id, peer FROM system.peers"))
            local = rows_to_list(s.execute("SELECT host_id, broadcast_address FROM system.local"))
>           peers_and_local.append(str(sorted(peers + local)))
E           TypeError: '<' not supported between instances of 'NoneType' and 'UUID'
```

Locally, I see also:
```
12:42:06,474 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2123 | Check peer table for node3 with ip address 127.0.49.33 : peers=[[UUID('fd396b2b-3162-4dc4-832d-fbe5777a2f14'), '127.0.49.2'], [UUID('eac15ed3-8f32-421b-91da-5d5bf5a0caa8'), '127.0.49.3'], [UUID('444777d5-1b55-4aa8-afd4-9d6b5f237fbd'), '127.0.49.1']] 
12:42:06,474 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2124 | Check peer table for node3 with ip address 127.0.49.33 : local=[[UUID('eac15ed3-8f32-421b-91da-5d5bf5a0caa8'), '127.0.49.33']] 
12:42:06,474 update_cluster_layout_tests WARNING update_cluster_layout_tests.py:2126 | Expected 2 peers but got 3
```

I'm not sure why this is seen consistently in debug mode and not in dev/release modes.
It's probably related to relative timing of racing async calls.

My analysis shows that `storage_service::on_join` is called with the old endpoint inet_address,
calling `on_change` with `application_state::STATUS` and `versioned_value::STATUS_NORMAL`,
calling `handle_state_normal(endpoint)`.

There log messages show:
```
WARN  2022-11-08 12:55:15,208 [shard 0] storage_service - Not updating host ID 81909e9f-202d-47ce-94da-3406dab51f26 for 127.0.63.3 because it's mine
```

BUT https://github.com/scylladb/scylladb/blob/a9063f9582f0ea22dd2c95d457c5626f23f0256e/service/storage_service.cc#L971-L977

calls `update_topology(endpoint, get_dc_rack_for(endpoint))`

since 421557b40aa8d1adf5016722418936a99595cb20

which puts it back into the topology.

And, https://github.com/scylladb/scylladb/blob/a9063f9582f0ea22dd2c95d457c5626f23f0256e/service/storage_service.cc#L997-L999
calls `notify_joined(endpoint)` which is not the root cause of this issue, but there's no reason to notify that the old endpoint joined when it owns no tokens and is marked for removal.

Later on, https://github.com/scylladb/scylladb/blob/a9063f9582f0ea22dd2c95d457c5626f23f0256e/service/storage_service.cc#L1217-L1218
Sees the old endpoint in token_metadata (after update_topology inserted it back)
and calls `do_update_system_peers_table(endpoint, state, value)` which adds it back to system.peers, after it was removed from it.

The following change fixes the issue for me locally:
```
diff --git a/service/storage_service.cc b/service/storage_service.cc
index 03611a7ed0..9896d4e6da 100644
--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -969,11 +969,15 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
     }
 
     bool is_member = tmptr->is_member(endpoint);
+    bool do_notify_joined = !is_member && !endpoints_to_remove.contains(endpoint);
+
     // Update pending ranges after update of normal tokens immediately to avoid
     // a race where natural endpoint was updated to contain node A, but A was
     // not yet removed from pending endpoints
-    if (!is_member) {
-        tmptr->update_topology(endpoint, get_dc_rack_for(endpoint));
+    if (do_notify_joined) {
+        auto dc_rack = get_dc_rack_for(endpoint);
+        slogger.debug("handle_state_normal: update_topology: endpoint={} dc={} rack={}", endpoint, dc_rack.dc, dc_rack.rack);
+        tmptr->update_topology(endpoint, std::move(dc_rack));
     }
     co_await tmptr->update_normal_tokens(owned_tokens, endpoint);
     co_await update_pending_ranges(tmptr, format("handle_state_normal {}", endpoint));
@@ -994,7 +998,7 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
     }
 
     // Send joined notification only when this node was not a member prior to this
-    if (!is_member) {
+    if (do_notify_joined) {
         co_await notify_joined(endpoint);
     }
 
```

### PR DESCRIPTION
**Pull request by bhalevy**
Currently, when replacing a node ip, keeping the old host,
we might end up with the the old endpoint in system.peers
if it is inserted back into the topology by `handle_state_normal`
when on_join is called with the old endpoint.

Then, later on, on_change sees that:
```
    if (get_token_metadata().is_member(endpoint)) {
        co_await do_update_system_peers_table(endpoint, state, value);
```

As described in #11925.

Fixes #11925

### CODE REVIEW

**Comment by asias** [comment]
@bhalevy  11903 is the wrong issue number

**Comment by bhalevy** [comment]
> @bhalevy 11903 is the wrong issue number

thanks, it should be #11925, will fix.

**Comment by bhalevy** [comment]
In ([a48f048](https://github.com/scylladb/scylladb/commit/a48f048a96a2451164e881a5fbbfe76a9f3f7dff)): fixed github issue number references

File: service/storage_service.cc [hunk 1]
```diff
@@ -969,10 +969,12 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
     }
 
     bool is_member = tmptr->is_member(endpoint);
+    bool do_notify_joined = !is_member && !endpoints_to_remove.contains(endpoint);
+
     // Update pending ranges after update of normal tokens immediately to avoid
     // a race where natural endpoint was updated to contain node A, but A was
     // not yet removed from pending endpoints
-    if (!is_member) {
+    if (do_notify_joined) {
```
Comment (xemul) [inline service/storage_service.cc:977]: This place worries me. The thing is that subsequent call to `update_normal_tokens` will assert if the endpoint will not be met in the topology. However, this `if (do_notify_joined)` is not necessarily in sync with this requirement.

Comment (bhalevy) [reply service/storage_service.cc:977]: I'll see if in all cases that the endpoint is in endpoints_to_remove its owned_tokens is empty.
If this is the case, we can skip both update_topology and update_normal_tokens (that returns immediately `if (tokens.empty())`)

**Comment by bhalevy** [comment]
In v3 ([38d8777](https://github.com/scylladb/scylladb/commit/38d8777d425ef9d6afbbdd35ef6d5380f9de1287)):
- call `update_normal_tokens` from `handle_state_normal` only if the endpoint is not in `endpoints_to_remove`
  - and verify that `owned_tokens` is not empty in this case, while if the endpoint is to be remove, owned_tokens IS empty.

**Comment by xemul** [comment]
LGTM, thanks!
@asias , please, re-review the updated first patch too

File: service/storage_service.cc [hunk 2]
```diff
@@ -969,13 +969,27 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
...
-    // Update pending ranges after update of normal tokens immediately to avoid
-    // a race where natural endpoint was updated to contain node A, but A was
-    // not yet removed from pending endpoints
-    if (!is_member) {
-        tmptr->update_topology(endpoint, get_dc_rack_for(endpoint));
+    bool do_notify_joined = false;
+
+    if (endpoints_to_remove.contains(endpoint)) [[unlikely]] {
+        if (!owned_tokens.empty()) {
+            on_fatal_internal_error(slogger, format("endpoint={} is marked for removal but still owns {} tokens", endpoint, owned_tokens.size()));
```
Comment (avikivity) [inline service/storage_service.cc:976]: Is the code above responsible for ensuring owned_tokens is empty? That is, we are asserting an function-local invariant here, not a node-scope or cluster-scope invariant?

It's much easier to be sure that a function-local invariant indeed holds compared to a node-scope or cluster-scope invariant.

Comment (bhalevy) [reply service/storage_service.cc:976]: Yes, `owned_tokens` is function local, and prepared in the code block right above this assertion.

**Comment by asias** [comment]
> LGTM, thanks! @asias , please, re-review the updated first patch too

Looks good to me.

**Pull request merged**